Update composer file for better semantic versioning

cgsmith opened the pull request:
I recommend committing the composer.lock file to the repository. Composer also recommends this since it is [not a library](https://getcomposer.org/doc/01-basic-usage.md#commit-your-composer-lock-file-to-version-control)

If it was a library you could still technically commit the lock file and composer would ignore it if the smarty-to-twig-converter was used as a dependency of a project. Since there are [no dependents](https://packagist.org/packages/oxid-esales/smarty-to-twig-converter/dependents?order_by=downloads) of this code it is okay to commit the lock file. 



Fixes [7758](https://bugs.oxid-esales.com/view.php?id=7758)

Commit:
Adds composer lock file to project

Commit:
Update to changelog
- Should use "Unreleased" without specifying a tag

Commit:
Fix a reference to old Twig docs

Sieg commented:
Hey @cgsmith, thanks for the pull request. We will discuss this topic in the team soon.

Regarding your second commit to the changelog and the change of "tag", I will not agree here - having the tag will improve the release experience, as at the moment of the change, you know how big is the change, and does it require minor, major or the patch is enough, so the person who is releasing, later, will not need to think and analyze the changes. Is there a better approach for solving this question?

cgsmith commented:
Thanks for checking @Sieg  :)

I don't think having a tag is appropriate until you can determine what is in the release. For example, adding a lock file could be considered a minor release and wouldn't warrant a major 2.0 or 3.0 release. I do agree with you that the release manager may not know what the change is. Other projects I've seen that get around this is by implementing something like this in the [GitHub PR template](https://github.com/yiisoft/yii2/blob/master/.github/PULL_REQUEST_TEMPLATE.md).

Sieg commented:
@cgsmith Regarding the PR template - its only handles the small portion of changes, and helpful for the person who merges the change, but again, this person can change the "unreleased" tag from patch->minor or major, if the change requires it, and currently we are doing this. So, if its written 2.0.0, means, someone's change already was big enough for the 2.0.0, but its not like - we don't know yet.

We also talked a bit about adding those composer.lock files today, and as I agree with you that this is not really a component, but more independent "project", its still can be installed on different versions of PHP, and different dependent component versions may be required and pulled by composer automatically without the lock. If we lock, then we will lock for one concrete case, as it would be with regular deployable project (you know what PHP you use there), but we don't have concrete case here.

cgsmith commented:
Maybe it is best not to have the `composer.lock` committed then but to adjust the `composer.json` to allow better semantic versioning?

Sieg commented:
@cgsmith, do you mean, to be more strict on the versions? Please describe the idea more.

cgsmith committed:
Revert composer.lock and update composer.json to align better to semver

cgsmith committed:
remove lock

cgsmith commented:
@Sieg I changed the PR to reflect the idea. I am curious what you think. I also added the 2.0.0 tag back into the changelog but also noted it with `[Unreleased]`.

I think the [type of library](https://getcomposer.org/doc/04-schema.md#type) matches the composer.json file best. I also think the caret versions are best with just the major.minor versions.

Thanks again for looking :)

Sieg commented:
Hey @cgsmith, yeah, for this case we can simplify those requirements to minors. Thanks for the improvements.
